### PR TITLE
Borders with rounded corners and drop shadows.

### PR DESCRIPTION
**Pull request by erikcorry**

### CODE REVIEW

File: src/style.toit [hunk 1]
```diff
@@ -205,6 +205,243 @@ class BorderWidth:
...
+  static get corner_radius/int -> RoundedCornerOpacity_:
+    cached := cache_.get corner_radius
+    if cached: return cached
+    new := RoundedCornerOpacity_.private_ corner_radius
+    cache_[corner_radius] = new
+    return new
+
+  static TABLE_SIZE_ ::= 256
+  // The heights of a top-right quarter circle of radius [TABLE_SIZE_].
+  static QUARTER_CIRCLE_ ::= create_quarter_circle_array_ TABLE_SIZE_
```
Comment (floitsch) [inline src/style.toit:395]: Type the static.
If it's not a constant, then the type-checker doesn't yet infer the type.

File: src/style.toit [hunk 2]
```diff
@@ -205,6 +205,243 @@ class BorderWidth:
...
+
+  static create_quarter_circle_array_ size:
```
Comment (floitsch) [inline src/style.toit:397]:
```suggestion
  static create_quarter_circle_array_ size -> ByteArray:
```

File: src/true_color.toit [hunk 3]
```diff
@@ -95,6 +96,41 @@ class Canvas_ extends Canvas:
...
+    // transparent.  We don't check for the case where 0 is opaque and 1 is
+    // transparent, because pngunzip fixes that for us.
+    if alpha[1] == 0xff and (zero_alpha == 0xff or zero_alpha == 0):
+      if zero_alpha == 0xff:
+        h := (pixels.size + source_line_stride - source_byte_width) / source_line_stride
+        // Draw the zeros.
+        rectangle x y --w=source_width --h=h --color=(BIG_ENDIAN.uint24 palette 0)
+      // Draw the ones.
+      transform.xyo x y 0: | x2 y2 o2 |
+        bitmap_draw_bitmap x2 y2 --color=palette[3] --orientation=o2 --source=pixels --source_width=source_width --source_line_stride=source_line_stride --destination=red_ --destination_width=width_ --bytewise
```
Comment (floitsch) [inline src/true_color.toit:119]: line is too long.

File: src/true_color.toit [hunk 4]
```diff
@@ -95,6 +96,41 @@ class Canvas_ extends Canvas:
...
+        bitmap_draw_bitmap x2 y2 --color=palette[4] --orientation=o2 --source=pixels --source_width=source_width --source_line_stride=source_line_stride --destination=green_ --destination_width=width_ --bytewise
```
Comment (floitsch) [inline src/true_color.toit:120]: line is too long.

File: src/true_color.toit [hunk 5]
```diff
@@ -95,6 +96,41 @@ class Canvas_ extends Canvas:
...
+        bitmap_draw_bitmap x2 y2 --color=palette[5] --orientation=o2 --source=pixels --source_width=source_width --source_line_stride=source_line_stride --destination=blue_ --destination_width=width_ --bytewise
```
Comment (floitsch) [inline src/true_color.toit:121]: line is too long.

File: src/true_color.toit [hunk 6]
```diff
@@ -95,6 +96,41 @@ class Canvas_ extends Canvas:
...
+      return
+    // Unfortunately one of the alpha values is not 0 or 0xff, so we can't use
+    // the bitmap draw primitive.  We can blow it up to bytes, then use the
+    // bitmap_draw_bytemap.
+    h := (pixels.size + source_line_stride - source_byte_width) / source_line_stride
+    bytemap := ByteArray source_width * h
+    bitmap_draw_bitmap 0 0 --color=1 --source=pixels --source_width=source_width --source_line_stride=source_line_stride --destination=bytemap --destination_width=source_width --bytewise
```
Comment (floitsch) [inline src/true_color.toit:128]: line too long.

File: src/true_color.toit [hunk 7]
```diff
@@ -95,6 +96,41 @@ class Canvas_ extends Canvas:
...
+    transform.xyo x y 0: | x2 y2 o2 |
+      bitmap_draw_bytemap x2 y2 --alpha=alpha --orientation=o2 --source=bytemap --source_width=source_width --palette=palette --destination=red_ --destination_width=width_
```
Comment (floitsch) [inline src/true_color.toit:130]: line too long.

File: src/true_color.toit [hunk 8]
```diff
@@ -95,6 +96,41 @@ class Canvas_ extends Canvas:
...
+      bitmap_draw_bytemap x2 y2 --alpha=alpha --orientation=o2 --source=bytemap --source_width=source_width --palette=palette[1..] --destination=green_ --destination_width=width_
```
Comment (floitsch) [inline src/true_color.toit:131]: line too long.

File: src/true_color.toit [hunk 9]
```diff
@@ -95,6 +96,41 @@ class Canvas_ extends Canvas:
...
+      bitmap_draw_bytemap x2 y2 --alpha=alpha --orientation=o2 --source=bytemap --source_width=source_width --palette=palette[2..] --destination=blue_ --destination_width=width_
```
Comment (floitsch) [inline src/true_color.toit:132]: line too long.

File: src/two_bit_texture.toit [hunk 10]
```diff
@@ -60,3 +60,31 @@ abstract class Canvas_ extends Canvas:
...
+    // transparent, because pngunzip fixes that for us.
+    if alpha[1] == 0xff and (zero_alpha == 0xff or zero_alpha == 0):
+      if zero_alpha == 0xff:
+        h := (pixels.size + source_line_stride - source_byte_width ) / source_line_stride
+        // Draw the zeros.
+        rectangle x y --w=source_width --h=h --color=(nearest_color_ palette 0)
+      // Draw the ones.
+      transform.xyo x y orientation: | x2 y2 o2 |
+        color := nearest_color_ palette 3
+        bitmap_draw_bitmap x2 y2 --color=(color & 1) --orientation=o2 --source=pixels --source_width=source_width --source_line_stride=source_line_stride --destination=plane_0_ --destination_width=width_
```
Comment (floitsch) [inline src/two_bit_texture.toit:87]: line too long.

File: src/two_bit_texture.toit [hunk 11]
```diff
@@ -60,3 +60,31 @@ abstract class Canvas_ extends Canvas:
...
+        bitmap_draw_bitmap x2 y2 --color=(color >> 1) --orientation=o2 --source=pixels --source_width=source_width --source_line_stride=source_line_stride --destination=plane_1_ --destination_width=width_
```
Comment (floitsch) [inline src/two_bit_texture.toit:88]: line too long.

File: src/two_color.toit [hunk 12]
```diff
@@ -66,3 +66,32 @@ class Canvas_ extends Canvas:
...
+    // transparent, because pngunzip fixes that for us.
+    if alpha[1] == 0xff and (zero_alpha == 0xff or zero_alpha == 0):
+      if zero_alpha == 0xff:
+        h := (pixels.size + source_line_stride - source_byte_width ) / source_line_stride
+        // Draw the zeros.
+        rectangle x y --w=source_width --h=h --color=(nearest_color_ palette 0)
+      // Draw the ones.
+      transform.xyo x y orientation: | x2 y2 o2 |
+        color := nearest_color_ palette 3
+        bitmap_draw_bitmap x2 y2 --color=(color & 1) --orientation=o2 --source=pixels --source_width=source_width --source_line_stride=source_line_stride --destination=pixels_ --destination_width=width_
```
Comment (floitsch) [inline src/two_color.toit:95]: line too long.

**Review by floitsch — APPROVED** [review]
LGTM.

**Review by erikcorry — COMMENTED** [review]
Put some small loops in true_color.toit to do components.

Shortened all the long lines that were mentioned.

**Pull request merged**